server/channels/app/cloud.go

```diff
@@ -266,3 +268,91 @@ func (a *App) SendSubscriptionHistoryEvent(userID string) (*model.SubscriptionHi
 	}
 	return a.Cloud().CreateOrUpdateSubscriptionHistoryEvent(userID, int(userCount))
 }
+
+func (a *App) DoSubscriptionRenewalCheck() {
+	if !a.License().IsCloud() || !a.Config().FeatureFlags.CloudAnnualRenewals {
+		return
+	}
+
+	subscription, err := a.Cloud().GetSubscription("")
+	if err != nil {
+		a.Log().Error("Error getting subscription", mlog.Err(err))
+		return
+	}
+
+	if subscription == nil {
+		a.Log().Error("Subscription not found")
+		return
+	}
+
+	sysVar, err := a.Srv().Store().System().GetByName(model.CloudRenewalEmail)
+	if err != nil {
+		// We only care about the error if it wasn't a not found error
+		if _, ok := err.(*store.ErrNotFound); !ok {
+			a.Log().Error(err.Error())
+		}
+	}
+
+	prevSentEmail := int64(0)
+	if sysVar != nil {
+		// We don't care about parse errors because it's possible the value is empty, and we've already defaulted to 0
+		prevSentEmail, _ = strconv.ParseInt(sysVar.Value, 10, 64)
+	}
+
+	if subscription.WillRenew == "true" {
+		// They've already completed the renewal process so no need to email them.
+		// We can zero out the system variable so that this process will work again next year
+		if prevSentEmail != 0 {
+			sysVar.Value = "0"
+			err = a.Srv().Store().System().SaveOrUpdate(sysVar)
+			if err != nil {
+				a.Log().Error("Error saving system variable", mlog.Err(err))
+			}
+		}
+		return
+	}
+
+	var emailFunc func(email, locale, siteURL string) error
+
+	daysToExpiration := subscription.DaysToExpiration()
+
+	// Only send the email if within the period and it's not already been sent
+	// This allows the email to send on day 59 if for whatever reason it was unable to on day 60
+	if daysToExpiration <= 60 && daysToExpiration > 30 && prevSentEmail != 60 {
+		emailFunc = a.Srv().EmailService.SendCloudRenewalEmail60
+		prevSentEmail = 60
+	} else if daysToExpiration <= 30 && daysToExpiration > 7 && prevSentEmail != 30 {
+		emailFunc = a.Srv().EmailService.SendCloudRenewalEmail30
+		prevSentEmail = 30
+	} else if daysToExpiration <= 7 && daysToExpiration > 3 && prevSentEmail != 7 {
+		emailFunc = a.Srv().EmailService.SendCloudRenewalEmail7
+		prevSentEmail = 7
+	}
+
+	if emailFunc == nil {
+		return
+	}
+
+	sysAdmins, aErr := a.getSysAdminsEmailRecipients()
+	if aErr != nil {
+		a.Log().Error("Error getting sys admins", mlog.Err(aErr))
+		return
+	}
+
+	for _, admin := range sysAdmins {
+		err = emailFunc(admin.Email, admin.Locale, *a.Config().ServiceSettings.SiteURL)
+		if err != nil {
+			a.Log().Error("Error sending renewal email", mlog.Err(err))
+		}
+	}
```

Comment:
Interesting, I would have expected that we would have had an `emailAdmins()` func laying around by now. (just thinking out loud 😄 )

More importantly, just want to check that it's reasonable to log the errors here and continue. In theory they could never get emails, but that would imply some larger server issue that doesn't really make sense to handle right?

Reply:
SMTP is pretty fire-and-forget, so yeah the only real error that would happen here is inability to actually contact the SMTP server, or an issue with the template engine itself.

It can get pretty complicated to ensure that every admin individually received the email - particularly around the mechanism to prevent resending (and the clearing mechanism so that the campaign can happen again next year). The list of admin's is fluid, so the list of who should receive could change day to day.

I've added a small change that will ensure that _at least one_ admin receives the email - if all of the sends fail, it will now try again the next day. I think this strikes a happy medium between safety and over engineering